stdlib/public/core/Reverse.swift

```diff
@@ -119,12 +132,69 @@ public struct ReverseCollection<
 
   public typealias _Element = Base.Iterator.Element
   public subscript(position: Index) -> Base.Iterator.Element {
-    return _base[_base.previous(position).base]
+    return _base[_base.previous(position.base)]
   }
 
   public let _base: Base
 }
 
+/// An index that traverses the same positions as an underlying index,
+/// with inverted traversal direction.
+public struct ReverseRandomAccessIndex<
+  Base : RandomAccessCollection
+> : Strideable {
+
+  /// The position corresponding to `self` in the underlying collection.
+  public let base: Base.Index
+
+  public typealias Stride = Base.Index.Stride
+
+  @warn_unused_result
+  public func distance(to other: ReverseRandomAccessIndex) -> Stride {
+    return base.distance(to: other.base)
+  }
+
+  @warn_unused_result
+  public func advanced(by n: Stride) -> ReverseRandomAccessIndex {
+    return ReverseRandomAccessIndex(base: base.advanced(by: n))
+  }
+}
+
+@warn_unused_result
+public func == <Base : RandomAccessCollection>(
+  lhs: ReverseRandomAccessIndex<Base>, rhs: ReverseRandomAccessIndex<Base>
+  ) -> Bool {
```

Comment:
I was going to look into doing that. I wanted to see what the code looked like and get a first round of feedback on it before doing that.